### PR TITLE
feat: add catboost ranker

### PR DESCRIPTION
**Pull request by bwanglzu**
Add CatBoost ranker to jina rankers.

### CODE REVIEW

File: jinahub/rankers/CatBoostRanker/catboost.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.q_features = query_features
+        self.d_features = document_features
+        self.label = label
+        self.model_path = model_path
+        self.catboost_parameters = catboost_parameters
+        self.model = None
+        if self.model_path and os.path.exists(self.model_path):
```
Comment (JoanFM) [inline jinahub/rankers/CatBoostRanker/catboost.py:34]: u can call `self.load` directly to enforce the design that restarting a Pea is like reloading the model

File: jinahub/rankers/CatBoostRanker/catboost.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            ]
+            for match in doc.matches:
+                d_feature_vector = [
+                    match.tags.get(docum_feature) for docum_feature in self.d_features
+                ]
+                feature_vectors.append(q_feature_vector + d_feature_vector)
+        feature_vectors = np.array(feature_vectors)
+        return Pool(data=feature_vectors, label=self.label, group_id=group_ids)
+
+    @requests(on='train')
```
Comment (JoanFM) [inline jinahub/rankers/CatBoostRanker/catboost.py:68]: maybe use `/train` to be in sync with `dump` and `load`?

Comment (bwanglzu) [reply jinahub/rankers/CatBoostRanker/catboost.py]: yes, updated

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                data=data,
+                label=label,
+                group_id=group_id,
+                weight=self._extract_weights(docs),
+            )
+        else:
+            return Pool(data=data, label=label, group_id=group_id)
+
+    @requests(on='/train')
+    def train(self, docs: DocumentArray, parameters: Optional[Dict] = {}, **kwargs):
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:101]: `docs` is Optional, `parameters` is not

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 4]
```diff
@@ -97,33 +97,36 @@ def build_catboost_pool(self, docs: DocumentArray):
...
+                'You need to train your model before make prediction, call `train` endpoint first.'
+            )
+        preds = self.model.predict(feature_vecs)
+        matches = docs.traverse_flat(traversal_paths=['m'])
+        for match, pred in zip(matches, preds):
+            match.tags[self.label] = pred
 
     @requests(on='/dump')
-    def dump(self, parameters: Dict, **kwargs):
+    def dump(self, parameters: Dict = {}, **kwargs):
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:121]: why do we need a default?

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 5]
```diff
@@ -97,33 +97,36 @@ def build_catboost_pool(self, docs: DocumentArray):
...
+    @requests(on='/rank')
+    def rank(self, docs: DocumentArray, parameters: Dict = {}, **kwargs):
+        feature_vecs, _, _ = self._extract_features(docs)
+        if not self.model.is_fitted():
+            raise ValueError(
+                'You need to train your model before make prediction, call `train` endpoint first.'
+            )
+        preds = self.model.predict(feature_vecs)
+        matches = docs.traverse_flat(traversal_paths=['m'])
+        for match, pred in zip(matches, preds):
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:117]: The objective of the ranker is to change the order of the matches which is not happening here. What catboost ranker should give is a score

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/rankers/CatboostRanker/README.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+f = Flow().add(uses='jinahub://CatboostRanker')
+
+da_to_train = DocumentArray()
+da_to_predict = DocumentArray()
+# note, to see how we build a document array, please refer to tests/conftest.py
+
+with f:
+    f.post(on='/train', inputs=da_to_train)
+    f.post(on='/predict', inputs=da_to_predict)
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/README.md:116]: is search the endpoint

File: jinahub/rankers/CatboostRanker/README.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+- `query_features` the tag names to extract value from query document. Each query-match pair will be combined into a feature vector.
+- `match_features` the tag names to extract value from match document. Each query-match pair will be combined into a feature vector.
+- `label` the tag name considered as groundtruth.
+- `weight` (Optional)the tag name to store the importance of query, stored in each query document.
+- `model_path` (Optional)the default model path to dump/load the document.
+- `catboost_parameters` (Optional) the parameters to train catboost ranker, please refer to catboost documentation for more info.
+
+### Inputs 
+
+`Document`, with `tags` and matches with `tags`.
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/README.md:132]: With tags corresponding to the features

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+from jina import DocumentArray, Executor, requests
+from jina.logging.logger import JinaLogger
+
+
+class CatboostRanker(Executor):
+    def __init__(
+        self,
+        query_features: List[str],
+        match_features: List[str],
+        label: str,
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:19]: I would call it `relevance_label`

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    ):
+        super().__init__(*args, **kwargs)
+        self.q_features = query_features
+        self.m_features = match_features
+        self.label = label
+        self.weight = weight
+        self.model_path = model_path
+        self.catboost_parameters = catboost_parameters
+        self.model = None
+        self.logger = JinaLogger('catboost logger')
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:40]:
```suggestion
        self.logger = JinaLogger(self.__class__.__name__)
```

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+                    match.tags.get(docum_feature) for docum_feature in self.m_features
+                ]
+                feature_vectors.append(q_feature_vector + m_feature_vector)
+        return np.array(feature_vectors), label_vector, group_ids
+
+    def _extract_weights(self, docs: DocumentArray):
+        """Extract weights data from docs, weights are needed if the query_docs has different
+        importance. Once given, we'll get values stored in the Document.tags based on the field name.
+        """
+        weight_vector = []
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:82]: can we do it in a list-comprehension way?

File: jinahub/rankers/CatboostRanker/catboost_ranker.py (hunk 9 above)
Comment (JoanFM) [reply jinahub/rankers/CatboostRanker/catboost_ranker.py:40]: Although u can remove as it is not used

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/rankers/CatboostRanker/catboost_ranker.py (hunk 10 above)
Comment (bwanglzu) [reply jinahub/rankers/CatboostRanker/catboost_ranker.py:82]: i don't see it brings anything, only reduce readability

File: jinahub/rankers/CatboostRanker/catboost_ranker.py (hunk 8 above)
Comment (bwanglzu) [reply jinahub/rankers/CatboostRanker/catboost_ranker.py:19]: i think lablel could be anything user want to set? such as `relevance`, or `groundtruth` ..

Comment (JoanFM) [reply jinahub/rankers/CatboostRanker/catboost_ranker.py:19]: no, groundtruth label wont make sense. The label is a relevance label for the prediction

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        :param docs: Documents to fit the model, each docs should have fields corresponded to
+          the :attr:`query_features` and has matches with corresponded :attr:`match_features` to
+          extract feature from. An additional field will be attached to the matched documents as was defined
+          as :attr:`label`.
+        :param parameters: Parameters to set, we'll get :attr:`catboost_parameters` from the parameters field.
+        """
+        feature_vecs, _, _ = self._extract_features(docs)
+        if not self.model.is_fitted():
+            raise ValueError(
+                'You need to train your model before make prediction, call `train` endpoint first.'
```
Comment (tadejsv) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:134]:
```suggestion
                'You need to train your model before making predictions, call the `train` endpoint first.'
```

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+from typing import Dict, List, Optional
+
+import numpy as np
+from catboost import CatBoostRanker, Pool
+
+from jina import DocumentArray, Executor, requests
+from jina.logging.logger import JinaLogger
+
+
+class CatboostRanker(Executor):
```
Comment (tadejsv) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:14]: You can add a docstring there

File: jinahub/rankers/CatboostRanker/tests/conftest.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        if 0.8 <= relevance <= 1:
+            brand = 5
+        elif 0.6 <= relevance < 0.8:
+            brand = 4
+        elif 0.4 <= relevance < 0.6:
+            brand = 3
+        elif 0.2 <= relevance < 0.4:
+            brand = 2
+        else:
+            brand = 1
```
Comment (tadejsv) [inline jinahub/rankers/CatboostRanker/tests/conftest.py:53]: To make it easier to read, I would replace all this with `np.digitize` (you can do it also outside the loop)

File: jinahub/rankers/CatboostRanker/README.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    f.post(on='/search', inputs=da_to_search)
+    f.post(on='/dump', parameters={'model_path': '/tmp/model.cbm'})
+    f.post(on='/load', parameters={'model_path': '/tmp/model.cbm'})
+```
+
+Parameters:
+
+- `query_features` the tag names to extract value from query document. Each query-match pair will be combined into a feature vector.
+- `match_features` the tag names to extract value from match document. Each query-match pair will be combined into a feature vector.
+- `label` the tag name considered as groundtruth.
```
Comment (JoanFM) [inline jinahub/rankers/CatboostRanker/README.md:125]: this is not accurate

File: jinahub/rankers/CatboostRanker/catboost_ranker.py (hunk 11 above)
Comment (JoanFM) [reply jinahub/rankers/CatboostRanker/catboost_ranker.py:134]: This is not true, U do not need to call `train` endpoint first. U need to provide a fitted model, even trained offline or calling training endpoint

File: jinahub/rankers/CatboostRanker/catboost_ranker.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+        :param parameters: Parameters pass to this endpoint, expect :attr:`model_path` to be set.
+        """
+        model_path = parameters.get('model_path', None)
+        if model_path:
+            self._save_model(model_path)
+        else:
+            raise ValueError('Please specify a `model_path` inside parameters variable')
+
+    @requests(on='/load')
+    def load(self, parameters: Dict, **kwargs):
```
Comment (tadejsv) [inline jinahub/rankers/CatboostRanker/catboost_ranker.py:156]: Do we need the `load` endpoint at all? I don't see why anyone would want to load a model after it has already been loaded for the first time

File: jinahub/rankers/CatboostRanker/tests/conftest.py (hunk 13 above)
Comment (bwanglzu) [reply jinahub/rankers/CatboostRanker/tests/conftest.py:53]: nice! learned something new @tadejsv

File: jinahub/rankers/CatboostRanker/catboost_ranker.py (hunk 15 above)
Comment (bwanglzu) [reply jinahub/rankers/CatboostRanker/catboost_ranker.py:168]: i'd like to add it, user might need to deploy multiple flows, dump here and load there etc.

Comment (tadejsv) [reply jinahub/rankers/CatboostRanker/catboost_ranker.py:168]: I see

**Review by tadejsv — APPROVED** [review]
Looks good to me now

**Review by JoanFM — APPROVED** [review]

**Comment by JoanFM** [comment]
Good job!

**Pull request merged**